exporter/metadataexporter/exporter.go

```diff
@@ -19,21 +34,385 @@ type tagValueCountFromDB struct {
+	switch datasource {
+	case pipeline.SignalTraces.String():
+		e.tracesTracker.AddValue(key, value)
+	}
+}
+
+func (e *metadataExporter) shouldSkipAttributeUVT(_ context.Context, key, datasource string) bool {
+	typ := e.getType(key, datasource)
+	var cnt int
+
+	switch datasource {
+	case pipeline.SignalTraces.String():
+		if _, ok := e.alwaysIncludeTracesAttributes[key]; ok {
+			return false
+		}
+		cnt = e.tracesTracker.GetUniqueValueCount(makeUVTKey(key, datasource))
+		if typ == "string" && e.cfg.MaxDistinctValues.Traces.MaxStringDistinctValues > 0 && cnt > int(e.cfg.MaxDistinctValues.Traces.MaxStringDistinctValues) {
+			return true
+		}
+		if typ == "float64" || typ == "int64" {
+			return true
+		}
+	}
+	return false
+}
+
+func (e *metadataExporter) filterAttrs(ctx context.Context, attrs map[string]any, datasource string) map[string]any {
+	filtered := make(map[string]any, len(attrs))
+	var skippedAttrs, nonSkipAttrs []string
+
+	maxLen := int(e.cfg.MaxDistinctValues.Traces.MaxStringLength)
+	for k, v := range attrs {
+		vStr := fmt.Sprint(v)
+		if len(vStr) == 0 || len(vStr) > maxLen {
+			skippedAttrs = append(skippedAttrs, k)
+			continue
+		}
+		e.addToUVT(ctx, makeUVTKey(k, datasource), vStr, datasource)
+		if !e.shouldSkipAttributeUVT(ctx, k, datasource) {
+			filtered[k] = v
+			nonSkipAttrs = append(nonSkipAttrs, k)
+		} else {
+			skippedAttrs = append(skippedAttrs, k)
+		}
+	}
+
+	e.set.Logger.Debug("filtered attributes",
+		zap.String("datasource", datasource),
+		zap.Int("skipped_count", len(skippedAttrs)),
+		zap.Strings("skipped_attributes", skippedAttrs),
+		zap.Strings("non_skipped_attributes", nonSkipAttrs),
+	)
+	return filtered
+}
+
+func makeFingerprintCacheKey(a, b uint64, datasource string) string {
+	builder := strings.Builder{}
+	builder.Grow(40 + len(datasource))
+	builder.WriteString(strconv.FormatUint(a, 10))
+	builder.WriteByte(':')
+	builder.WriteString(strconv.FormatUint(b, 10))
+	builder.WriteByte(':')
+	builder.WriteString(datasource)
+	return builder.String()
+}
+
+func (e *metadataExporter) writeToStmt(_ context.Context, stmt driver.Batch, ds pipeline.Signal, resourceFingerprint, fprint uint64, rAttrs, filtered map[string]any, roundedUnixMilli int64) (bool, error) {
+	cacheKey := makeFingerprintCacheKey(fprint, uint64(roundedUnixMilli), ds.String())
+	if item := e.fingerprintCache.Get(cacheKey); item != nil && item.Value() {
+		return true, nil
+	}
+
+	if err := stmt.Append(
+		roundedUnixMilli,
+		ds,
+		resourceFingerprint,
+		fprint,
+		flattenJSONToStringMap(rAttrs),
+		flattenJSONToStringMap(filtered),
+	); err != nil {
+		return false, err
+	}
+
+	e.fingerprintCache.Set(cacheKey, true, ttlcache.DefaultTTL)
+	return false, nil
+}
+
 func (e *metadataExporter) PushTraces(ctx context.Context, td ptrace.Traces) error {
+	stmt, err := e.conn.PrepareBatch(ctx, "INSERT INTO signoz_metadata.distributed_attributes_metadata", driver.WithReleaseConnection())
+	if err != nil {
+		return err
+	}
```

Comment:
add defer functions which aborts the statement ?

Reply:
I read about it and don't really understand why it's required. Please help me understand why we explicitly need to abort?

Reply:
I tried checking it and didn't find anything concrete, though some of their test does contain the abort statement for batches.